### PR TITLE
Rewrites things for Guzzle

### PR DESCRIPTION
**Pull request by israelshirk**
Also adds Travis-CI support.  Downside is that PHPUnit 5.5 requires PHP >= 5.6; actually running it on 5.5 works fine.  Unit tests show basic guzzle tests where the remote is mocked out - I haven't gotten to adding production data to tests there yet, but it's pretty straightforwards.

### CODE REVIEW

**Comment by markroland** [comment]
I am traveling this week and will look at merging this next week. Thanks!

File: README.md
Comment (markroland) [inline README.md]: I have created a Travis CI page for this at https://travis-ci.org/markroland/converge-api-php -- let's use that.

File: composer.json
Comment (markroland) [inline composer.json]: I'd like to remove this. I prefer to have versions represented as tags. That's Composer's recommendation: https://getcomposer.org/doc/04-schema.md#version

File: phpunit.xml
Comment (markroland) [inline phpunit.xml]: Also, I prefer this to be named as phpunit.xml.dist as it was before. That makes it a configuration specific to this package.

The <filter> block with the whitelist needs to be added back in order to generate code coverage reports.

File: src/ConvergeApi.php
Comment (markroland) [inline src/ConvergeApi.php:73]: Why have you removed the HTTP Method parameter? What happens if there's an API method that requires an HTTP method other than POST?

**Review by markroland — CHANGES_REQUESTED** [review]

**Review by markroland — APPROVED** [review]

**Review by markroland — APPROVED** [review]
Can you add a specific example of what this fixes? I haven't looked at this in a while :)

**Review by markroland — APPROVED** [review]

**Comment by markroland** [comment]
Thanks for your contributions, I really appreciate them!

**Comment by markroland** [comment]
Hi Israel, are you able to make the updates I requested so I can merge this? If not, I'm willing to make the updates and merge.

**Comment by israelshirk** [comment]
### Hey!

Sorry, I was crazy busy for a while there.

### GET/POST

I didn't see any GET requests in the developer guide (except for one instance where it was optional between GET and POST) so I figured it just simplified things.  I can re-add as desired; the nice thing about switching to just POST is that the various functions (`ccsale`, `txnquery`, etc) can be expressed by just implementing __call() and validating that the method exists using a list of the known converge methods, then just shipping the call off - so no coding necessary for each individual method.

### Version

Definitely.  Will have this in in a moment.

### Use Case

My use case is now that I can do E2E testing of a transaction-processing service which makes several calls to Converge and other internal services, mocking out all of the services' responses like this:

#### Caller

And the wrapper around the call to ConvergeApi looks like this to inject the handler:

``` php

    private function _processor()
    {
        $convergeApi = app(
            \markroland\Converge\ConvergeApi::class, 
            array(
                $this->credentials['merchantId'],
                $this->credentials['userId'],
                $this->credentials['pin'],
                in_array($this->credentials['paymentGatewayEnvironment'], ['PROD', 'live'])
            )
        );

        if (app()->bound('ConvergeGuzzleHandler')) {
            $convergeApi->setHandler(app('ConvergeGuzzleHandler'));
        }

        return $convergeApi;
    }

```

#### Unit Test (Laravel)

``` php
// Other services are excluded here

$expectedResponse = implode("\n", array(
    'ssl_approval_code=lsdakjf',
    'ssl_cvv2_response=',
    'ssl_exp_date=1222',
    'ssl_result_message=APPROVAL',
    'ssl_invoice_number=SALKDFJASLDKFJ-ASLDFJKLASDKFJ',
    'ssl_amount=200',
    'ssl_transaction_currency=USD',
    'ssl_result=0',
    'ssl_txn_id=LASKDJF-SALDKFJ-ASLDKFJASLD',
    'ssl_card_number=41**********1111',
    'ssl_txn_time=11/19/2013 12:02:29 AM'
));

$expectedRequestProperties = array(
    'cited' => 'meta-first_name_cited meta-last_name_cited',
    'payer' => 'Jimmy Fallon',
    "ssl_amount" => "523.75",
    "ssl_exp_date" => "1223",
    "ssl_avs_zip" => "83706",
    "ssl_avs_address" => "123 Test St",
    "ssl_first_name" => "Jimmy",
    "ssl_last_name" => "Fallon",
    "ssl_card_number" => "4111111111111111",
    "ssl_cvv2cvc2" => "123",
    "ssl_get_token" => "N",
    "ssl_transaction_type" => "ccsale",
    "ssl_merchant_id" => "11111",
    "ssl_user_id" => "webpage",
    "ssl_pin" => "PINPIN",
    "ssl_show_form" => "false",
    "ssl_result_format" => "ascii",
    "ssl_test_mode" => "false",
    "ssl_city" => "Boise",
    "ssl_state" => "ID",
    "ssl_salestax" => "0",
    "ssl_description" => "Payment made through nolatepayments.com",
    "ssl_receipt_link_method" => "GET",
    "ssl_cvv2cvc2_indicator" => "1",
);

// historyContainer here is passed by reference.  Later
// we can inspect it to see history.
$historyContainer = [];

// This is kind of convoluted but gets it done.  Welcome
// to modularization without convenience methods :)
$historyMiddleware = Middleware::history($historyContainer);
$mockBody = Psr7\stream_for($expectedResponse);
$mockResponse = new Response(200, [], $mockBody);
$mockHandler = new MockHandler([$mockResponse]);
$handlerStack = HandlerStack::create($mockHandler);
$handlerStack->push($historyMiddleware);

app()->instance('ConvergeGuzzleHandler', $handlerStack);

$this->postSignedJson(
    '/transaction/process/charge',
    json_decode($this->exampleData)
)->seeJsonEquals($expectedResult);

$this->assertSame(count($historyContainer), 1);

$httpRequest = $historyContainer[0]['request'];
$httpResponse = $historyContainer[0]['response'];
$body = (string)$httpRequest->getBody();
parse_str($body, $requestProperties);

$this->assertEquals($expectedRequestProperties, $requestProperties);
$this->assertEquals((string)$httpRequest->getUri(), $expectedUri);
```

**Comment by markroland** [comment]
Thanks for the quick reply! This came back to my attention because I got a call from Converge and also another PR related to the "besha2ready" conversion (http://www.besha2ready.com/converge#convergeapi) so I'm going to merge that as well.

**Comment by israelshirk** [comment]
Sweet!

**Review by markroland — APPROVED** [review]
Thanks!

**Comment by markroland** [comment]
I'm okay with leaving the method calls as POSTs. Just a few. more things before I merge:

1) I need the "whitelist" section of the phpunit.xml.dist put back because I need that for generating code coverage reports.

2) Can you explain the rewrite you've done on the ascii response kvp splitting? 

3) Just an opinion question, do you think this warrants a major version update from 0 to 1 or just a minor version update. I'm not sure what I think. It's probably mature enough it could be on a "version 1" now and the change from curl to guzzle is pretty big.

Thanks again.

**Comment by israelshirk** [comment]
1. Re-added

2. When receiving 500 errors or anything of that sort there would be an HTML response with multiple (or no) `=` signs in some lines in the response.  This caused a notice storm and filled up logging.  To some extent this solution glosses over the problem, but makes it usable for the moment if you were to have incorrect credentials or similar.

3. I think I'd bump the minor version with this; supporting all methods via a magic function call would definitely justify a v1.0.  But it's not at all unreasonable, and that could go into a 2.0.  I've got another PR in the pipeline for the txnquery request FYI.

I'll fix tests for the new API endpoint and then everything should be passing again.

**Comment by israelshirk** [comment]
Ok, looks good after rebase and fixing API URLs.

**Comment by israelshirk** [comment]
Sweet, thanks!

**Pull request merged**
